maven_iuvs/echelle.py

```diff
@@ -0,0 +1,1292 @@
+    else: 
+        return dark_options[0] # TO DO: Make this more intelligent
+
+
+def find_dark_options(input_light_idx, idx_list_to_search):
+    """
+    Looks for darks matching the observation described by input_light_idx.
+
+    Parameters
+    ----------
+    input_light_idx : dictionary
+                      a dictionary entry of metadata for some observation
+    idx_list_to_search : list of dictionaries
+                         where each dictionary is the metadata for dark files
+
+    Returns
+    ----------
+    dark_options : list of dictionaries
+                   where each dictionary is the metadata for dark files that 
+                   match input_light_idx
+    """
+    if not ech_islight(input_light_idx):
+        raise ValueError('Input file index corresponds to a dark observation, cannot find matching dark.')
+    
+    half_orbit = datetime.timedelta(hours=2)
+    dark_options = [didx for didx in idx_list_to_search 
+                    if (np.abs(didx['datetime'] - input_light_idx['datetime']) < half_orbit
+                        and didx['binning'] == input_light_idx['binning']
+                        # and didx['mcp_gain'] == input_light_idx['mcp_gain']
+                        and didx['int_time'] == input_light_idx['int_time']
+                        # and iuvs_orbno_from_fname(didx['name']) == iuvs_orbno_from_fname(input_light_idx['name'])
+                        and iuvs_segment_from_fname(didx['name']) == iuvs_segment_from_fname(input_light_idx['name'])
+                        and ech_isdark(didx))]
+    
+    return dark_options
+
+
+def ech_isdark(fidx):
+    """
+    Identifies whether an echelle file contains dark integrations by checking the gain.
+
+    Parameters
+    ----------
+    fidx : dictionary
+           a single dictionary entry of metadata for some observation
+    Returns
+    ----------
+    True or False
+    """
+
+    return 'dark' in fidx['name']
+
+
+def ech_islight(fidx):
+    """
+    Identifies whether an echelle file has light (observation) integrations.
+
+    Parameters
+    ----------
+    fidx : dictionary
+           a single dictionary entry of metadata for some observation
+    Returns
+    ----------
+    True or False
+    """
+    return not ech_isdark(fidx)
+
+
+def make_dark_index(ech_l1a_idx):
+    """
+    Takes the index of l1a file metadata, ech_l1a_idx, and makes a similar index that will be used 
+    to find dark files. note that the return value is NOT darks only.
+    
+    Parameters
+    ----------
+    ech_l1a_idx : list of dictionaries
+                  metadata for all light observation files.
+    """
+    dark_idx = [fidx for fidx in ech_l1a_idx if (('orbit' in fidx['name']) and ech_isdark(fidx))]
+    dark_idx = sorted(dark_idx, key=lambda i:i['datetime'])
+    
+    return dark_idx
+
+# Count rates ----------------------------------------------------------
+
+
+def get_avg_pixel_count_rate(hdul, spapixrange, spepixrange, return_npix=True):
```

Comment:
probably this should be a generic routine in a different file?

Reply:
I didn't write this one or put it in here, but happy to move it. How about moving it to geometry?

Reply:
it probably belongs in 'integration' if it belongs anywhere; 'geometry' is specifically routines for handling spacecraft and pointing location information.